c10/core/TensorImpl.h

```diff
@@ -1092,21 +1134,27 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
     refresh_contiguous();
   }
 
+  // Inference tensor doesn't have version counter,
+  // set_version_counter is no-op for them.
   void set_version_counter(
     const c10::VariableVersion& version_counter) noexcept {
-    version_counter_ = version_counter;
+    if (!this->is_inference_tensor()) {
+      version_counter_ = version_counter;
+    }
   }
 
   void set_version_counter(
     c10::VariableVersion&& version_counter) noexcept {
```

Comment:
Since `set_version_counter` may now no longer move out of the rvalue reference here, it is probably clearer to replace this with a pass-by-value signature `set_version_counter(c10::VariableVersion version_counter)`. Otherwise, a call `set_version_counter(std::move(version_counter))` might imply that version counter is dead after the call, but if this is an inference tensor, it's NOT dead, and that would be very surprising to a code reader.